clients/cobol-lsp-vscode-extension/src/services/ControlFlowService.ts

```diff
@@ -103,75 +127,152 @@ class AnalysisTask {
 
 export class ControlFlowAnalysisService implements AnalysisServiceDelegate {
   private tasks: Map<string, AnalysisTask>;
-  private callbacks: Map<string, ControlFlowAnalysisCallback>;
+  private latestResults: Map<string, LatestResultData>;
   private diagnosticService: DiagnosticService;
+  private requestVersion: number = 0;
 
   public constructor(
     private mainChannel?: vscode.OutputChannel,
     private logChannel?: vscode.LogOutputChannel,
   ) {
     this.tasks = new Map<string, AnalysisTask>();
-    this.callbacks = new Map<string, ControlFlowAnalysisCallback>();
     this.diagnosticService = new DiagnosticService();
+    this.latestResults = new Map<string, LatestResultData>();
+    this.requestVersion = 1;
+  }
+
+  public async invalidate(documentUri: string, rejectPromise: boolean) {
+    this.logChannel?.debug(`Invalidate document: ${documentUri}`);
+
+    const latestResult = this.latestResults.get(documentUri);
+    if (latestResult) {
+      if (rejectPromise) {
+        this.latestResults.delete(documentUri);
+        latestResult.reject("invalidate");
+      } else {
+        latestResult.requestVersion = 0;
+      }
+    }
+    const task = this.tasks.get(documentUri);
+    if (task) {
+      this.logChannel?.debug(`Stop task for the document: ${documentUri}`);
+      this.tasks.delete(documentUri);
+      await task.abort();
+    }
+  }
+
+  public async getAnalysis(documentUri: string): Promise<AnalysisResult> {
+    this.logChannel?.debug(
+      `Get analysis request for the document: ${documentUri}`,
+    );
+
+    const latestResult = this.latestResults.get(documentUri);
+    if (latestResult) {
+      return latestResult.promise;
+    } else {
+      return this.createLatestResultPromise(documentUri, 0);
+    }
+  }
+
+  public makeControlFlowAstNotificationHandler() {
+    return (result: ApiResult) => {
+      this.handleControlFlowAst(result).catch(() => {});
+    };
   }
 
-  public queueAnalysis(programs: Program[], documentUri: string) {
+  async handleControlFlowAst(result: ApiResult) {
+    this.logChannel?.debug("Handle AST from backend");
+    if (result.documentUri) {
+      await this.invalidate(result.documentUri, false);
+      if (result.controlFlowAST.length > 0) {
+        this.queueAnalysis(result.controlFlowAST, result.documentUri);
+      }
```

Comment:
Just to make sure there is not reordering due to changes/work termination.
```suggestion
      const p = this.invalidate(result.documentUri, false);
      if (result.controlFlowAST.length > 0) {
        this.queueAnalysis(result.controlFlowAST, result.documentUri);
      }
      await p;
```
It might look less confusing if the `invalidate` is not async and `await task.abort();` is replaced with just `return task.abort();`, or the method is split into two private implementations `invalidateImpl` (non-async) and `task_terminate` (can be async).

Reply:
Got it, fixed